osquery/tables/system/darwin/extended_attributes.cpp

```diff
@@ -0,0 +1,223 @@
+  std::string attribute_data;
+  int return_value;
+  int buffer_length;
+};
+
+// Foward declares of all parse functions
+void parseWhereFrom(QueryData &results,
+                    const XAttrAttribute &x_att,
+                    const std::string &path,
+                    const std::string &directory);
+void parseQuarantineFile(QueryData &results,
+                         const XAttrAttribute &x_att,
+                         const std::string &path,
+                         const std::string &directory);
+
+// Extended attribute name to the required parse function. If there is no parse
+// function, the data is returned as is.
+const std::map<std::string,
+               void (*)(QueryData &,
+                        const XAttrAttribute &,
+                        const std::string &,
+                        const std::string &)> xParseMap = {
+    {"com.apple.metadata:kMDItemWhereFroms", parseWhereFrom},
+    {"com.apple.quarantine", parseQuarantineFile},
+};
+
+// Handle any errors thrown by sys/xattr
+std::string handleError() {
+  switch (errno) {
+  case ENOATTR:
+    return "No such attribute";
+  case ENOTSUP:
+    return "No system support, or support disabled";
+  case ERANGE:
+    return "Size too small to hold attribute";
+  case EPERM:
+    return "The named attribute is not permitted for this type of object";
+  case EINVAL:
+    return "Name is invalid, or options has an unsupported bit set";
+  case EISDIR:
+    return "The path is not a regular file";
+  case ENOTDIR:
+    return "Part of the path was not a directory";
+  case ENAMETOOLONG:
+    return "The filename was too long";
+  case EACCES:
+    return "Insufficient permissions to read the requested file";
+  case ELOOP:
+    return "Too many symbolic links were encountered";
+  case EFAULT:
+    return "The path or name is invalid";
+  case EIO:
+    return "There was an I/O error while reading";
+  default:
+    return "No data";
+  }
+}
+
+// Pull the requested extended attribute from the path and return
+// the XAttrAttribute structure
+struct XAttrAttribute getAttribute(const std::string &path,
+                                   const std::string attribute) {
+  struct XAttrAttribute x_att;
+  x_att.buffer_length =
+      getxattr(path.c_str(), attribute.c_str(), NULL, 0, 0, 0);
+  char *buffer = (char *)malloc(x_att.buffer_length);
+  x_att.return_value = getxattr(path.c_str(), attribute.c_str(), buffer,
+                                x_att.buffer_length, 0, 0);
+
+  if (x_att.return_value != -1) {
+    x_att.attribute_data = std::string(buffer, x_att.buffer_length);
+  } else {
+    x_att.attribute_data = std::string("");
+    x_att.buffer_length = 0;
+  }
+  free(buffer);
+  return x_att;
+}
+
+// Pull the list of all the extended attributes for a path
+std::vector<std::string> parseExtendedAttributeList(const std::string &path) {
+  std::vector<std::string> attributes;
+  ssize_t value = listxattr(path.c_str(), 0x0, 0x0, 0x0);
```

Comment:
weird 0x0, what does this mean? should you be using a `nullptr`, either way, let's be consistent with the rest of the code in the file.